<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Building &amp; loading datastores · ReadDatastores.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.png" alt="ReadDatastores.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit">ReadDatastores.jl</span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../read-datastores/">Datastore types</a></li><li class="is-active"><a class="tocitem" href>Building &amp; loading datastores</a><ul class="internal"><li class="toplevel"><a class="tocitem" href="#Loading-Read-Datastores-1"><span>Loading Read Datastores</span></a></li><li><a class="tocitem" href="#Using-the-macros-and-string-literals-1"><span>Using the macros and string literals</span></a></li></ul></li><li><a class="tocitem" href="../indexing/">Indexing &amp; Iteration</a></li><li><a class="tocitem" href="../additional-methods/">Additional Methods</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Building &amp; loading datastores</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Building &amp; loading datastores</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/BioJulia/ReadDatastores.jl/blob/master/docs/src/build-datastores.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Building-Read-Datastores-1"><a class="docs-heading-anchor" href="#Building-Read-Datastores-1">Building Read Datastores</a><a class="docs-heading-anchor-permalink" href="#Building-Read-Datastores-1" title="Permalink"></a></h1><p>To build a read datastore you first need to decide what sort of read data you have.</p><p>The process of building these datastores is consistent, but for each datastore there are datastore specific options the constructor will accept:</p><article class="docstring"><header><a class="docstring-binding" id="ReadDatastores.PairedReads" href="#ReadDatastores.PairedReads"><code>ReadDatastores.PairedReads</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">PairedReads{A}(rdrx::FASTQ.Reader, rdry::FASTQ.Reader, outfile::String, name::Union{String,Symbol}, minsize::Integer, maxsize::Integer, fragsize::Integer, orientation::PairedReadOrientation) where {A&lt;:DNAAlphabet}</code></pre><p>Construct a Paired Read Datastore from a pair of FASTQ file readers.</p><p>Paired-end sequencing reads typically come in the form of two FASTQ files, often named according to a convention <code>*_R1.fastq</code> and <code>*_R2.fastq</code>. One file contains all the &quot;left&quot; sequences of each pair, and the other contains all the &quot;right&quot; sequences of each pair. The first read pair is made of the first record in each file.</p><p><strong>Arguments</strong></p><ul><li><code>rdrx::FASTQ.Reader</code>: The reader of the <code>*_R1.fastq</code> file.</li><li><code>rdxy::FASTQ.Reader</code>: The reader of the <code>*_R2.fastq</code> file.</li><li><code>outfile::String</code>: A prefix for the datastore&#39;s filename, the full filename will include a &quot;.prseq&quot; extension, which will be added automatically.</li><li><code>name::String</code>: A string denoting a default name for your datastore. Naming datastores is useful for downstream applications.</li><li><code>minsize::Integer</code>: A minimum read length (in base pairs). When building the datastore, if any pair of reads has one or both reads shorter than this cutoff, then the pair will be discarded.</li><li><code>maxsize::Integer</code>: A maximum read length (in base pairs). When building the datastore, if any read has a greater length, it will be resized to this maximum length and added to the datastore.</li><li><code>fragsize::Integer</code>: The average fragment length of the paired end library that was sequenced. This value is entirely optional, but may be important for downstream applications.</li><li><code>orientation::PairedReadOrientation</code>: The orientation of the reads. Set it to <code>FwRv</code> for building a datastore from a sequenced paired end library, and set it to <code>RvFw</code> if you are building the datastore from reads sequenced from a long mate pair library.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl">julia&gt; using FASTX, ReadDatastores

julia&gt; fwq = open(FASTQ.Reader, &quot;test/ecoli_tester_R1.fastq&quot;)
FASTX.FASTQ.Reader{TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}}(BioGenerics.Automa.State{TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}}(TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}(&lt;mode=idle&gt;), 1, 1, false), nothing)

julia&gt; rvq = open(FASTQ.Reader, &quot;test/ecoli_tester_R2.fastq&quot;)
FASTX.FASTQ.Reader{TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}}(BioGenerics.Automa.State{TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}}(TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}(&lt;mode=idle&gt;), 1, 1, false), nothing)

julia&gt; ds = PairedReads{DNAAlphabet{2}}(fwq, rvq, &quot;ecoli-test-paired&quot;, &quot;my-ecoli-test&quot;, 250, 300, 0, FwRv)
[ Info: Building paired read datastore from FASTQ files
[ Info: Writing paired reads to datastore
[ Info: Done writing paired read sequences to datastore
[ Info: 0 read pairs were discarded due to a too short sequence
[ Info: 14 reads were truncated to 300 base pairs
[ Info: Created paired sequence datastore with 10 sequence pairs
Paired Read Datastore &#39;my-ecoli-test&#39;: 20 reads (10 pairs)
</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/ReadDatastores.jl/blob/b1713a5cecec0940ee73ccb87150a98f57f7d539/src/paired-reads.jl#L43-L95">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ReadDatastores.LongReads" href="#ReadDatastores.LongReads"><code>ReadDatastores.LongReads</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">LongReads{A}(rdr::FASTQ.Reader, outfile::String, name::String, min_size::Integer) where {A&lt;:DNAAlphabet}</code></pre><p>Construct a Long Read Datastore from a FASTQ file reader.</p><p><strong>Arguments</strong></p><ul><li><code>rdr::FASTQ.Reader</code>: The reader of the fastq formatted file.</li><li><code>outfile::String</code>: A prefix for the datastore&#39;s filename, the full filename will include a &quot;.loseq&quot; extension, which will be added automatically.</li><li><code>name::String</code>: A string denoting a default name for your datastore. Naming datastores is useful for downstream applications.</li><li><code>min_size::Integer</code>: A minimum read length (in base pairs). When building the datastore, if any read sequence is shorter than this cutoff, then the read will be discarded.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl">julia&gt; using FASTX, ReadDatastores

julia&gt; longrdr = open(FASTQ.Reader, &quot;test/human_nanopore_tester_2D.fastq&quot;)
FASTX.FASTQ.Reader{TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}}(BioGenerics.Automa.State{TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}}(TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}(&lt;mode=idle&gt;), 1, 1, false), nothing)

julia&gt; ds = LongReads{DNAAlphabet{2}}(longrdr, &quot;human-nanopore-tester&quot;, &quot;nanopore-test&quot;, 0)
[ Info: Building long read datastore from FASTQ file
[ Info: Writing long reads to datastore
[ Info: Done writing paired read sequences to datastore
[ Info: 0 reads were discarded due to a too short sequence
[ Info: Writing index to datastore
[ Info: Built long read datastore with 10 reads
Long Read Datastore &#39;nanopore-test&#39;: 10 reads
</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/ReadDatastores.jl/blob/b1713a5cecec0940ee73ccb87150a98f57f7d539/src/long-reads.jl#L49-L81">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ReadDatastores.LinkedReads" href="#ReadDatastores.LinkedReads"><code>ReadDatastores.LinkedReads</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia">LinkedReads{A}(fwq::FASTQ.Reader, rvq::FASTQ.Reader, outfile::String, name::String, format::LinkedReadsFormat, max_read_len::Integer, chunksize::Int = 1000000) where {A&lt;:DNAAlphabet}</code></pre><p>Construct a Linked Read Datastore from a pair of FASTQ file readers.</p><p>Paired sequencing reads typically come in the form of two FASTQ files, often named according to a convention <code>*_R1.fastq</code> and <code>*_R2.fastq</code>. One file contains all the &quot;left&quot; sequences of each pair, and the other contains all the &quot;right&quot; sequences of each pair. The first read pair is made of the first record in each file.</p><p><strong>Arguments</strong></p><ul><li><code>rdrx::FASTQ.Reader</code>: The reader of the <code>*_R1.fastq</code> file.</li><li><code>rdxy::FASTQ.Reader</code>: The reader of the <code>*_R2.fastq</code> file.</li><li><code>outfile::String</code>: A prefix for the datastore&#39;s filename, the full filename will include a &quot;.prseq&quot; extension, which will be added automatically.</li><li><code>name::String</code>: A string denoting a default name for your datastore. Naming datastores is useful for downstream applications.</li><li><code>format::LinkedReadsFormat</code>: Specify the linked reads format of your fastq files. If you have plain 10x files, set format to <code>Raw10x</code>. If you have 10x reads output in the UCDavis format, set the format to <code>UCDavis10x</code>.</li><li><code>chunksize::Int = 1000000</code>: How many read pairs to process per disk batch during the tag sorting step of construction.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl">julia&gt; using FASTX, ReadDatastores

julia&gt; fqa = open(FASTQ.Reader, &quot;test/10x_tester_R1.fastq&quot;)
FASTX.FASTQ.Reader{TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}}(BioGenerics.Automa.State{TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}}(TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}(&lt;mode=idle&gt;), 1, 1, false), nothing)

julia&gt; fqb = open(FASTQ.Reader, &quot;test/10x_tester_R2.fastq&quot;)
FASTX.FASTQ.Reader{TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}}(BioGenerics.Automa.State{TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}}(TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}(&lt;mode=idle&gt;), 1, 1, false), nothing)

julia&gt; ds = LinkedReads{DNAAlphabet{2s}}(fqa, fqb, &quot;10xtest&quot;, &quot;ucdavis-test&quot;, UCDavis10x, 250)
[ Info: Building tag sorted chunks of 1000000 pairs
[ Info: Dumping 83 tag-sorted read pairs to chunk 0
[ Info: Dumped
[ Info: Processed 83 read pairs so far
[ Info: Finished building tag sorted chunks
[ Info: Performing merge from disk
[ Info: Leaving space for 83 read_tag entries
[ Info: Chunk 0 is finished
[ Info: Finished merge from disk
[ Info: Writing 83 read tag entries
[ Info: Created linked sequence datastore with 83 sequence pairs
Linked Read Datastore &#39;ucdavis-test&#39;: 166 reads (83 pairs)
</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/ReadDatastores.jl/blob/b1713a5cecec0940ee73ccb87150a98f57f7d539/src/linked-reads.jl#L61-L110">source</a></section></article><h1 id="Loading-Read-Datastores-1"><a class="docs-heading-anchor" href="#Loading-Read-Datastores-1">Loading Read Datastores</a><a class="docs-heading-anchor-permalink" href="#Loading-Read-Datastores-1" title="Permalink"></a></h1><p>Once you have built a datastore, you can open it in other projects again using a <code>Base.open</code> method:</p><pre><code class="language-julia-repl">julia&gt; ds = open(PairedReads{DNAAlphabet{2}}, &quot;ecoli-test-paired.prseq&quot;, &quot;my-ecoli-pe&quot;)
Paired Read Datastore &#39;my-ecoli-pe&#39;: 20 reads (10 pairs)</code></pre><p>The open method takes a <code>ReadDatastore</code> type, the filename of the datastore, and, optionally a name for the datastore, if you omit the name, the datastore will use the default name that was specified on construction.</p><p>You can also use <code>do</code> block notation when opening a datastore and that will ensure that the underlying stream of the reader will close.</p><h2 id="Using-the-macros-and-string-literals-1"><a class="docs-heading-anchor" href="#Using-the-macros-and-string-literals-1">Using the macros and string literals</a><a class="docs-heading-anchor-permalink" href="#Using-the-macros-and-string-literals-1" title="Permalink"></a></h2><p>If you try to open a datastore with the <code>open</code> method as above, and you provide the wrong type you will get an error.</p><p>This will protect you from trying to open a long read datastore as a paired read datastore and such, but it&#39;s not always convenient.</p><p>For example, if you have a paired read datastore storing reads in 2-bit format and you tried to open it as a <code>PairedReads{DNAAlphabet{4}}</code> type you will still get an error.</p><p>This is obviously correct behaviour, you don&#39;t want to be loading sequences using a different encoding to the one they were saved with!</p><p>However, in a practical setting this will get annoying: maybe you want to use some long reads you put in a datastore a while ago but don&#39;t remember if your datastore file is a <code>LongReads{DNAAlphabet{4}}</code> or a <code>LongReads{DNAAlphabet{2}}</code>. Or maybe you get a <code>somefile.prseq</code> file from a colleague, and from the extension, you deduce it is paired reads but even then that&#39;s not guaranteed.</p><p>To solve this problem a few convenience macros are provided for you, so you can load datastores without specifying the datastore type, yet still avoid type uncertainty in your generated code.</p><p>The macro <code>@openreads</code> accepts a filepath, and optionally a datastore name. The macro is evaluated and expanded before you julia code is compiled. During that time, the header of the datastore file is peeked at, and the correct ReadDatastore subtype is deduced, and a correctly typed <code>open</code> statement is generated.</p><p>For example if the file <code>myreads.prseq</code> was a 2-bit encoded paired read datastore, and you had the following statement in your script: <code>ds = @openreads &quot;myreads.prseq&quot;</code></p><p>The statement would be expanded to: <code>ds = open(PairedReads{DNAAlphabet{2}}, &quot;myreads.prseq&quot;)</code></p><p>You can also open a datastore using a string literal e.g. <code>reads&quot;myreads.prseq&quot;</code>. When you do this, type type of the datastore is detected as with <code>@openreads</code>, however, rather than returning the expression <code>ds = open(PairedReads{DNAAlphabet{2}}, &quot;myreads.prseq&quot;)</code>, as <code>@openreads</code> does, the <code>open</code> is executed and the macro returns the value of <code>ds</code>.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>In order for <code>@openreads</code> or the literals to work properly in any script, the datastore file must exist prior to running the script. This is unavoidable because macros are evaluated and expanded first before the resulting expanded code is compiled and run.</p><p>So creating a datastore file and loading it again using <code>@openreads</code> within the same script will not work, and <code>@openreads</code> will try to peek at the file and deduce its contents before the script can generate the datastore file. You will get an error telling you the file can&#39;t be found / does not exist.</p><p>In an interactive setting, in which statements are entered, compiled and run by the REPL one at a time, this should rarely be a problem.</p></div></div><p><strong>In summary: If in doubt about the datastore type of a file, simply use <code>@openreads</code></strong></p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../read-datastores/">« Datastore types</a><a class="docs-footer-nextpage" href="../indexing/">Indexing &amp; Iteration »</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Saturday 6 June 2020 23:30">Saturday 6 June 2020</span>. Using Julia version 1.4.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
