<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Building &amp; loading datastores · ReadDatastores.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link href="../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../"><img class="logo" src="../assets/logo.png" alt="ReadDatastores.jl logo"/></a><h1>ReadDatastores.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../">Home</a></li><li class="current"><a class="toctext" href>Building &amp; loading datastores</a><ul class="internal"><li class="toplevel"><a class="toctext" href="#Loading-Read-Datastores-1">Loading Read Datastores</a></li></ul></li><li><a class="toctext" href="../indexing/">Indexing &amp; Iteration</a></li></ul></nav><article id="docs"><header><nav><ul><li><a href>Building &amp; loading datastores</a></li></ul><a class="edit-page" href="https://github.com/BioJulia/ReadDatastores.jl/blob/master/docs/src/build-datastores.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Building &amp; loading datastores</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Building-Read-Datastores-1" href="#Building-Read-Datastores-1">Building Read Datastores</a></h1><p>To build a read datastore you first need to decide what sort of read data you have.</p><p>If you have short (somewhat regular in length) paired reads such as that were sequenced from a paired-end library or a long mate pair library, you should build a <code>PairedReads</code> datastore.</p><p>If you have long reads of inconsistent length, such as you would get from single-molecule real-time sequencing of a long read library, you should build a <code>LongReads</code> datastore.</p><p>If you have short (somewhat regular in length) paired reads, and thr pairs are linked linked through some proximity tagging mechanism (like 10x), then you should build a LinkedReads datastore.</p><p>The process of building these datastores is consistent, but for each datastore there are datastore specific options the constructor will accept:</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ReadDatastores.PairedReads" href="#ReadDatastores.PairedReads"><code>ReadDatastores.PairedReads</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-julia">PairedReads(rdrx::FASTQ.Reader, rdry::FASTQ.Reader, outfile::String, name::String, minsize::Integer, maxsize::Integer, fragsize::Integer, orientation::PairedReadOrientation)</code></pre><p>Construct a Paired Read Datastore from a pair of FASTQ file readers.</p><p>Paired-end sequencing reads typically come in the form of two FASTQ files, often named according to a convention <code>*_R1.fastq</code> and <code>*_R2.fastq</code>. One file contains all the &quot;left&quot; sequences of each pair, and the other contains all the &quot;right&quot; sequences of each pair. The first read pair is made of the first record in each file.</p><p><strong>Arguments</strong></p><ul><li><code>rdrx::FASTQ.Reader</code>: The reader of the <code>*_R1.fastq</code> file.</li><li><code>rdxy::FASTQ.Reader</code>: The reader of the <code>*_R2.fastq</code> file.</li><li><code>outfile::String</code>: A prefix for the datastore&#39;s filename, the full filename will include a &quot;.prseq&quot; extension, which will be added automatically.</li><li><code>name::String</code>: A string denoting a default name for your datastore. Naming datastores is useful for downstream applications.</li><li><code>minsize::Integer</code>: A minimum read length (in base pairs). When building the datastore, if any pair of reads has one or both reads shorter than this cutoff, then the pair will be discarded.</li><li><code>maxsize::Integer</code>: A maximum read length (in base pairs). When building the datastore, if any read has a greater length, it will be resized to this maximum length and added to the datastore.</li><li><code>fragsize::Integer</code>: The average fragment length of the paired end library that was sequenced. This value is entirely optional, but may be important for downstream applications.</li><li><code>orientation::PairedReadOrientation</code>: The orientation of the reads. Set it to <code>FwRv</code> for building a datastore from a sequenced paired end library, and set it to <code>RvFw</code> if you are building the datastore from reads sequenced from a long mate pair library.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl">julia&gt; using FASTX, ReadDatastores

julia&gt; fwq = open(FASTQ.Reader, &quot;test/ecoli_tester_R1.fastq&quot;)
FASTX.FASTQ.Reader{TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}}(BioGenerics.Automa.State{TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}}(TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}(&lt;mode=idle&gt;), 1, 1, false), nothing)

julia&gt; rvq = open(FASTQ.Reader, &quot;test/ecoli_tester_R2.fastq&quot;)
FASTX.FASTQ.Reader{TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}}(BioGenerics.Automa.State{TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}}(TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}(&lt;mode=idle&gt;), 1, 1, false), nothing)

julia&gt; ds = PairedReads(fwq, rvq, &quot;ecoli-test-paired&quot;, &quot;my-ecoli-test&quot;, 250, 300, 0, FwRv)
[ Info: Building paired read datastore from FASTQ files
[ Info: Writing paired reads to datastore
[ Info: Done writing paired read sequences to datastore
[ Info: 0 read pairs were discarded due to a too short sequence
[ Info: 14 reads were truncated to 300 base pairs
[ Info: Created paired sequence datastore with 10 sequence pairs
Paired Read Datastore &#39;my-ecoli-test&#39;: 20 reads (10 pairs)
</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/ReadDatastores.jl/blob/aad3978e286d6ee97679dd3cbeb6ad8e89dae8c3/src/paired-reads.jl#L39-L91">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ReadDatastores.LongReads" href="#ReadDatastores.LongReads"><code>ReadDatastores.LongReads</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-julia">LongReads(rdr::FASTQ.Reader, outfile::String, name::String, min_size::Integer)</code></pre><p>Construct a Long Read Datastore from a FASTQ file reader.</p><p><strong>Arguments</strong></p><ul><li><code>rdr::FASTQ.Reader</code>: The reader of the fastq formatted file.</li><li><code>outfile::String</code>: A prefix for the datastore&#39;s filename, the full filename will include a &quot;.loseq&quot; extension, which will be added automatically.</li><li><code>name::String</code>: A string denoting a default name for your datastore. Naming datastores is useful for downstream applications.</li><li><code>min_size::Integer</code>: A minimum read length (in base pairs). When building the datastore, if any read sequence is shorter than this cutoff, then the read will be discarded.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl">julia&gt; using FASTX, ReadDatastores

julia&gt; longrdr = open(FASTQ.Reader, &quot;test/human_nanopore_tester_2D.fastq&quot;)
FASTX.FASTQ.Reader{TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}}(BioGenerics.Automa.State{TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}}(TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}(&lt;mode=idle&gt;), 1, 1, false), nothing)

julia&gt; ds = LongReads(longrdr, &quot;human-nanopore-tester&quot;, &quot;nanopore-test&quot;, 0)
[ Info: Building long read datastore from FASTQ file
[ Info: Writing long reads to datastore
[ Info: Done writing paired read sequences to datastore
[ Info: 0 reads were discarded due to a too short sequence
[ Info: Writing index to datastore
[ Info: Built long read datastore with 10 reads
Long Read Datastore &#39;nanopore-test&#39;: 10 reads
</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/ReadDatastores.jl/blob/aad3978e286d6ee97679dd3cbeb6ad8e89dae8c3/src/long-reads.jl#L33-L65">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ReadDatastores.LinkedReads" href="#ReadDatastores.LinkedReads"><code>ReadDatastores.LinkedReads</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-julia">LinkedReads(fwq::FASTQ.Reader, rvq::FASTQ.Reader, outfile::String, name::String, format::LinkedReadsFormat, readsize::Integer, chunksize::Int = 1000000)</code></pre><p>Construct a Linked Read Datastore from a pair of FASTQ file readers.</p><p>Paired sequencing reads typically come in the form of two FASTQ files, often named according to a convention <code>*_R1.fastq</code> and <code>*_R2.fastq</code>. One file contains all the &quot;left&quot; sequences of each pair, and the other contains all the &quot;right&quot; sequences of each pair. The first read pair is made of the first record in each file.</p><p><strong>Arguments</strong></p><ul><li><code>rdrx::FASTQ.Reader</code>: The reader of the <code>*_R1.fastq</code> file.</li><li><code>rdxy::FASTQ.Reader</code>: The reader of the <code>*_R2.fastq</code> file.</li><li><code>outfile::String</code>: A prefix for the datastore&#39;s filename, the full filename will include a &quot;.prseq&quot; extension, which will be added automatically.</li><li><code>name::String</code>: A string denoting a default name for your datastore. Naming datastores is useful for downstream applications.</li><li><code>format::LinkedReadsFormat</code>: Specify the linked reads format of your fastq files. If you have plain 10x files, set format to <code>Raw10x</code>. If you have 10x reads output in the UCDavis format, set the format to <code>UCDavis10x</code>.</li><li><code>chunksize::Int = 1000000</code>: How many read pairs to process per disk batch during the tag sorting step of construction.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl">julia&gt; using FASTX, ReadDatastores

julia&gt; fqa = open(FASTQ.Reader, &quot;test/10x_tester_R1.fastq&quot;)
FASTX.FASTQ.Reader{TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}}(BioGenerics.Automa.State{TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}}(TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}(&lt;mode=idle&gt;), 1, 1, false), nothing)

julia&gt; fqb = open(FASTQ.Reader, &quot;test/10x_tester_R2.fastq&quot;)
FASTX.FASTQ.Reader{TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}}(BioGenerics.Automa.State{TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}}(TranscodingStreams.TranscodingStream{TranscodingStreams.Noop,IOStream}(&lt;mode=idle&gt;), 1, 1, false), nothing)

julia&gt; ds = LinkedReads(fqa, fqb, &quot;10xtest&quot;, &quot;ucdavis-test&quot;, UCDavis10x, 250)
[ Info: Building tag sorted chunks of 1000000 pairs
[ Info: Dumping 83 tag-sorted read pairs to chunk 0
[ Info: Dumped
[ Info: Processed 83 read pairs so far
[ Info: Finished building tag sorted chunks
[ Info: Performing merge from disk
[ Info: Leaving space for 83 read_tag entries
[ Info: Chunk 1 is finished
[ Info: Finished merge from disk
[ Info: Writing down 83 read tag entries
[ Info: Created linked sequence datastore with 83 sequence pairs
Linked Read Datastore &#39;ucdavis-test&#39;: 166 reads (83 pairs)
</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/BioJulia/ReadDatastores.jl/blob/aad3978e286d6ee97679dd3cbeb6ad8e89dae8c3/src/linked-reads.jl#L62-L111">source</a></section><h1><a class="nav-anchor" id="Loading-Read-Datastores-1" href="#Loading-Read-Datastores-1">Loading Read Datastores</a></h1><p>Once you have built a datastore, you can open it in other projects again using a <code>Base.open</code> method:</p><pre><code class="language-julia-repl">julia&gt; ds = open(PairedReads, &quot;test/ecoli-pe.prseq&quot;, &quot;my-ecoli-pe&quot;)
Paired Read Datastore &#39;my-ecoli-pe&#39;: 20 reads (10 pairs)</code></pre><p>The open method takes a <code>ReadDatastore</code> type, the filename of the datastore, and, optionally a name for the datastore, which may be omitted, in which case the datastore will use the default name that was specified on construction. </p><footer><hr/><a class="previous" href="../"><span class="direction">Previous</span><span class="title">Home</span></a><a class="next" href="../indexing/"><span class="direction">Next</span><span class="title">Indexing &amp; Iteration</span></a></footer></article></body></html>
